rfcs/2020-03-17-2064-event-driven-observability.md

```diff
@@ -0,0 +1,275 @@
+
+## Sales Pitch
+
+This approach effectively separates concerns, keeping instrumentation points
+concise while giving us a dedicated place to make the data consistent, thorough,
+and discoverable.
+
+More specifically:
+
+* Using simple structs as the API leverages the compiler to enforce a consistent
+  set of keys and values everywhere an event is emitted.
+
+* Removing metrics derivation from the normal flow of code encourages
+  contributors to collect more thorough data.
+
+* Putting all events in a single module makes them easily to audit for
+  consistency.
+
+* The `InternalEvent` trait gives us a place to enforce uniform implementation
+  within events.
+
+* Being a thin layer over the existing `tracing` and `metrics` APIs should allow
+  most, if not all, of this additional code to be inlined or optimized away,
+  leaving us with the same excellent performance as the underlying libraries.
+
+## Drawbacks
+
+The primary downside of this approach is that it is unfamiliar to most
+developers and adds an additional step to the development process. We would need
+to take steps in documentation, API design, and code review to reduce friction
+for contributors as much as possible.
+
+## Alternatives
+
+The most realistic alternative is to simply not build this additional layer and
+use the normal `tracing` and `metrics` APIs. This has the benefit of familiarity
+and ease of use, but also brings all the downsides discussed above.
+
+Another alternative is to build tooling to help enforce consistency, provide
+discoverability, etc while using the normal APIs. This would be challenging to
+do in a precise way, and would not provide the benefits of separating concerns
+we'd get with explicit events. Since the convention would not be enforced by the
+compiler, the feedback loop would also likely be longer.
+
+## Outstanding Questions
+
+### Specific vs Reusable Events
+
+Consider the difference between the `FileEventReceived` event in the example
+above and something like the following:
+
+```rust
+pub struct EventReceived {
+    pub source_type: &'static str,
+    pub byte_size: usize,
+}
+```
+
+Or even:
+
+```rust
+pub struct EventProcessed {
+    pub component_kind: &'static str,
+    pub component_type: &'static str,
+    pub byte_size: usize,
+}
+```
+
+On one hand, more specific events like `FileEventReceived` can be tightly
+tailored to a specific callsite with data like `file`. On the other, they have
+a higher cognitive cost since they need to be implemented over and over for
+different sources, etc. This also makes consistency slightly more of
+a challenge.
+
+The more generic events can be quickly, easily, and consistently used across
+different components, but can be more verbose and include less contextual
+information. Extending them to allow for more dynamic data would defeat much of
+the purpose of using structs in the first place.
+
+Finding the right middle ground here is something that will likely come with
```

Comment:
Exactly, this will be something that we'll refine over time. And it'll never be perfect, but it should start to feel more automatic as head down this road. Speaking from experience, it took a couple of passes to settle on the right middle ground, afterward, it was mostly automatic. I plan to help help with this process in a separate PR where we can all discuss the tradeoffs.